File: src/types/ServiceCatalog.ts
export type ServiceCatalogBranch = 'Design' | 'Supervision'

export interface ServiceCatalogActivity {
  id: string
  name: string
  fixedCost: number
}

// branch determines billing behavior (see backend migration 0059): Design
// services are one-time fees; the single Supervision service's activities
// are monthly recurring fees, day-prorated for partial calendar months.
export interface ServiceCatalogItem {
  id: string
  name: string
  branch: ServiceCatalogBranch
  activities: ServiceCatalogActivity[]
}

// A single leaf pick made in ServicePickerDialog -- one activity, under one
// service, with the price it was selected at. Kept flat (rather than
// nested service -> activities) because every consumer (line items, quote
// prefill, contract scope) wants to iterate picks, not re-walk a tree.
// Design branch only -- see SelectedSupervisionActivity (Project.ts) for
// the Supervision equivalent, which also carries its own start/end dates.
export type SelectedActivityStatus = 'Not Started' | 'In Progress' | 'Complete' | 'Cancelled'

export interface SelectedServiceActivity {
  serviceId: string
  serviceName: string
  activityId: string
  activityName: string
  fixedCost: number
  // Present once this pick has actually been persisted to a project
  // (ProjectOut.selectedActivities) -- absent while it's still just a
  // fresh selection in ServicePickerDialog before the project is
  // created. id is this row's own identity (not activityId, the
  // catalog's display id) -- what Task.selectedActivityId and the
  // close/reopen actions operate on.
  id?: string
  status?: SelectedActivityStatus
  closedAt?: string
}

// An admin-configured "this Design activity must be Complete before
// this Supervision activity is eligible" rule -- see
// project_service._recompute_supervision_eligibility. Same shape as
// PermitPrerequisite (types/PermitCatalog.ts).
export interface SupervisionPrerequisite {
  id: string
  designActivityId: string
  designActivityName: string
  serviceName: string
}

File: backend/app/schemas/service_catalog.py
from typing import Literal

from pydantic import BaseModel, Field, condecimal


class ServiceCatalogActivityOut(BaseModel):
    id: str
    name: str
    fixedCost: float

    @staticmethod
    def from_model(activity) -> "ServiceCatalogActivityOut":
        return ServiceCatalogActivityOut(
            id=f"ACT-{activity.id:03d}",
            name=activity.name,
            fixedCost=float(activity.fixed_cost),
        )


class ServiceCatalogItemOut(BaseModel):
    id: str
    name: str
    branch: Literal["Design", "Supervision"]
    activities: list[ServiceCatalogActivityOut]

    @staticmethod
    def from_model(service) -> "ServiceCatalogItemOut":
        return ServiceCatalogItemOut(
            id=f"SVC-{service.id:03d}",
            name=service.name,
            branch=service.branch,
            activities=[ServiceCatalogActivityOut.from_model(a) for a in service.activities],
        )


class ServiceCatalogItemCreate(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    branch: Literal["Design", "Supervision"] = "Design"


class ServiceCatalogItemUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=150)


class ServiceCatalogActivityCreate(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    fixedCost: condecimal(ge=0, max_digits=12, decimal_places=2) = Field(default=0)  # type: ignore[valid-type]


class ServiceCatalogActivityUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    fixedCost: condecimal(ge=0, max_digits=12, decimal_places=2) | None = None  # type: ignore[valid-type]


class SupervisionPrerequisiteOut(BaseModel):
    """One admin-configured "this Design activity must be Complete
    before this Supervision activity is eligible" rule (migration
    0074) -- see project_service._recompute_supervision_eligibility.
    Same shape as PermitPrerequisiteOut."""

    id: str
    designActivityId: str
    designActivityName: str
    serviceName: str

    @staticmethod
    def from_model(prerequisite) -> "SupervisionPrerequisiteOut":
        activity = prerequisite.design_activity
        return SupervisionPrerequisiteOut(
            id=str(prerequisite.id),
            designActivityId=f"ACT-{activity.id:03d}",
            designActivityName=activity.name,
            serviceName=activity.service.name,
        )


class SupervisionPrerequisiteCreate(BaseModel):
    designActivityId: str = Field(min_length=1, max_length=20)
